README.rst
.. This README is meant for consumption by humans and pypi. Pypi can render rst files so please do not use Sphinx features.
   If you want to learn more about writing documentation, please check out: http://docs.plone.org/about/documentation_styleguide.html
   This text does not appear on pypi or github. It is a comment.

==============================================================================
uwosh.oie.studyabroadstudent
==============================================================================

Content types, workflows, and utilities for the Office of
International Education at the University of Wisconsin Oshkosh
(http://uwosh.edu/oie).


Features
========

- supports the submission, approval, publication, and management of study away
  programs
- supports the submission, approval, publication, and management of student
  applications for programs
- scripts to export legacy data from the first OIE workflow site (circa 2006,
  Archetypes-based, Plone 2.5) and import them to Plone 5 (Dexterity-based)



Documentation
=============

Full documentation for end users can be found in the "docs" folder.


Translations
============

None (yet)


Installation
============

Install uwosh.oie.studyabroadstudent by adding it to your buildout::

    [buildout]

    ...

    eggs =
        uwosh.oie.studyabroadstudent


and then running ``bin/buildout``


Export / Import
===============

This explains how to export data from the legacy site then import it into a new
site.

One-Time Install on Legacy Site
-------------------------------

Install these on the legacy site (you need to do this only once):

- Using the Management Interface, e.g. http://localhost:8080/OIE/manage_main,
  navigate to `portal_skins/custom` and create a Script (Python) called
  `listApplicationIDs` containing the code in the file
  `src/uwosh/oie/studyabroadstudent/scripts/listApplicationIDs.py`. This will
  be used to output the IDs of all applications on the site.

- Using the Management Interface, create a Script (Python) in
  `portal_skins/custom` called `extractApplication` containing the code in the
  file `src/uwosh/oie/studyabroadstudent/scripts/extractApplication.py` and add
  `id` to the Script (Python)'s parameter list. This will be used to export the
  application for a given ID.


Get the List of IDs to Export
-----------------------------

Each time you want to run the export, do this on the legacy site:

- Run the Script (Python) `listApplicationIDs` by clicking on the Test tab (you
  may first have to click on `listApplicationIDs` in the breadcrumbs before
  clicking on the Test tab). This outputs Python code in your browser window.
  This code is an array of application IDs.

- Save the output to a local file with the `.py` Python filename extension,
  e.g. `listApplicationIDsoutput.py`

- Copy `listApplicationIDsoutput.py` to the new server, using `scp`.

- On the new server, move `listApplicationIDsoutput.py` to
  `src/uwosh/oie/studyabroadstudent/scripts/listApplicationIDsoutput.py` so that
  the list of IDs can be loaded in the export script.

Choosing to Use a Public URL or an SSH Tunnel
---------------------------------------------

Ensure that, from the new site server, the legacy site is accessible.The
extract script uses XML-RPC to ask the legacy site to export each application.

You can specify the URL of the legacy site as HTTP or HTTPS.

You can choose to use the legacy site's public URL or you can ssh tunnel to the
legacy site server and specify a URL such as `http://localhost:8084:OIE`. The
advantage of using an ssh tunnel is that you can start a reserved ZEO client on
the legacy server and use it to do the export, rather than using the legacy
site's public URL and using a public ZEO client, which would affect any users
of the legacy site.

If you choose to use an ssh tunnel (recommended!), you must start the ssh tunnel
before you can proceed with the export steps below::

    ssh mylegacyserversshuser@mylegacyserver -L8084:localhost:8084

This opens an SSH tunnel on port 8084 so you can start a ZEO client that listens
on port 8084::

    cd /opt/Plone/zeocluster/
    bin/client_reserved start

You must keep this SSH tunnel open for the duration of the export process.

Export the Applications
-----------------------

On the new site server:

- Change directory to the Plone installation::

    cd /opt/Plone/zeocluster

- Run the script `extractApplicationsRemotely.py` from a command line, saving
  its output to a file::

      bin/instance run extractApplicationsRemotely.py > extractoutput.out

  This will take hours (to extract 8600 student applications it took over 3.5
  hours over the internet).

- You can override default values with `--http-ok=yes`, `--remote-user`,
  `--remote-password`, `--remote-server`, or you can use environment variables
  `REMOTEUSER`, `REMOTEPASSWD`, `REMOTESERVER` to specify those values::

   export REMOTEUSER=mylegacyploneuser
   export REMOTEPASSWD=mylegacyplonepassword
   export REMOTESERVER=localhost:8084/OIE
   bin/instance run ./src/uwosh/oie/studyabroadstudent/scripts/extractApplicationsRemotely.py --http-ok=yes > extractoutput.out

  If you use an SSH tunnel, be sure to specify `--http-ok=yes`, otherwise the
  script defaults to assuming HTTPS.

- If `extractApplicationsRemotely.py` did not finish running and extracted only
  a subset of the IDs, you can rerun it to skip the IDs it had
  previously read.

  First, you must generate the list of IDs that have already
  been exported, then you must move it to where the code expects to find it::

    grep '^[[]' extractoutput.out | awk -F \' 'BEGIN {print "already_read = ["} {printf "\"%s\",\n", $2} END { print "]\n" }' > already_read.py
    mv already_ready.py src/uwosh/oie/studentstudyabroad/

  Then you are ready to rerun the script::

    bin/instance run ./src/uwosh/oie/studyabroadstudent/scripts/extractApplicationsRemotely.py --http-ok=yes --skip-ids=yes --id-file=extractoutput.out > newextractoutput.out

- If you ran the script more than once and have more than one output file,
  combine all the output files into one file::

    cat extractoutput.out newextractoutput.out > combinedextractoutput.out

Convert the Export File to Python
---------------------------------

- Massage the output file to turn it into a Python file::

    src/uwosh/oie/studyabroadstudent/scripts/pythonify_extract.sh extractoutput.out > extractoutput.py

Import the Applications
-----------------------

- Run the script
  `src/uwosh/oie/studyabroadstudent/scripts/importApplications.py` to import
  the applications locally into the site, where the
  `uwosh.oie.studentstudyabroad` add-on must already be installed. By default,
  the import will try to create applications in the folder `applications` of
  the site `OIE` but you can override those values with `--site-id=Plone` and
  `--folder-id=another-folder-id`::

    bin/instance run src/uwosh/oie/studyabroadstudent/scripts/importApplications.py

  This took almost one hour on a MacBook Air.


Handling Data Problems
----------------------

If there is a problem with the data, such as a bad birth date year, you will be
dropped into a PDB prompt where you can inspect the values and make any changes
needed before continuing::


    (Pdb) l
    329  	        try:
    330  	            dateOfBirth=date(int((DateOfBirth_year is not None) and DateOfBirth_year or '1900'), month_values[((DateOfBirth_month is not None) and (DateOfBirth_month != '-- choose one --')) and DateOfBirth_month or 'January'], int(((DateOfBirth_day is not None) and (DateOfBirth_day != '-- choose one --') and DateOfBirth_day or 1)))
    331  	        except:
    332  	            import pdb;pdb.set_trace()
    333
    334  ->	        try:
    335  	            passportExpDate=date(int((PassportExpDate_year is not None and PassportExpDate_year is not '') and PassportExpDate_year or '1900'), month_values[(PassportExpDate_month == '-- choose one --' or PassportExpDate_month == '') and 'January' or PassportExpDate_month], int((PassportExpDate_day == '-- choose one --' or PassportExpDate_day == '') and 1 or PassportExpDate_day))
    336  	        except:
    337  	            import pdb;pdb.set_trace()
    338
    339  	        obj = api.content.create(
    (Pdb) DateOfBirth_year
    19991
    (Pdb) DateOfBirth_month
    'January'
    (Pdb) DateOfBirth_day
    '8'
    (Pdb) DateOfBirth_year=1991
    (Pdb) dateOfBirth=date(int((DateOfBirth_year is not None) and DateOfBirth_year or '1900'), month_values[((DateOfBirth_month is not None) and (DateOfBirth_month != '-- choose one --')) and DateOfBirth_month or 'January'], int(((DateOfBirth_day is not None) and (DateOfBirth_day != '-- choose one --') and DateOfBirth_day or 1)))
    (Pdb) dateOfBirth
    datetime.date(1991, 1, 8)
    (Pdb) id
    'oiestuapp_howlettj1287342001'
    (Pdb) c

Using the ID of the application that had the error, you can go to the site,
browse to that application (it will be at a URL like
`https://app.oie.uwosh.edu/Members/howlettj12/oiestuapp_howlettj1287342001`)
and edit and save it to correct the data error and prevent this particular data
error in future extracts and imports.

Rerunning the Import Multiple Times
-----------------------------------

The argument `--skip-existing` tells the script to check first if there is
already an existing application object with the same ID before creating one.

Verifying the Import
--------------------

To verify the import, on the new site:

- Using the Management Interface, create a Script (Python) in
  `portal_skins/custom` called `extractApplication` containing the code in the
  file `src/uwosh/oie/studyabroadstudent/scripts/extractApplication.py` and add
  `id` to the parameter list

- Create and run the script `extractApplicationsLocally.py` and save its output
  to a file, like this::

   bin/instance run extractApplicationsLocally.py > extractlocallyoutput.out

- Compare the contents of that file to that of the one you created remotely
  before::

   diff extractoutput.out extractlocallyoutput.out

Legacy Time Zones
=================

Since Plone 2.5, the time zones database has changed: the time zones `GMT-5`
and `GMT-6` have since been renamed `Etc/GMT+5` and `Etc/GMT+6` (see
https://community.plone.org/t/unknowntimezoneerror-pytz-quirks/4255/4 for why
the sign change). Some DateTime values in legacy data use the old time zone
designations, which causes an error in the `unpickler()` method in the file
`tzinfo.py`, part of the `pytz-2015.7-py2.7`.egg. This error prevents viewing
of legacy (migrated) `OIEStudentApplication` objects.

For the moment, the only way to get around this error is to patch the
`unpickler()` method in `tzinfo.py` and add the following lines right after
line 525 (the prophetic comment `Raises a KeyError if zone no longer exists,
which should never happen and would be a bug.`)::

    # Raises a KeyError if zone no longer exists, which should never happen
    # and would be a bug.
    newzone = zone
    if zone.find('GMT-') != -1:
        newzone = zone.replace('GMT-','Etc/GMT+')
    if zone.find('GMT+') != -1:
        import pdb;pdb.set_trace()
        newzone = zone.replace('GMT+','Etc/GMT-')
    if zone == 'GMT':
        newzone = 'Etc/GMT'
    if zone != newzone:
        logger.warn('fixing nonexistent timezone %s to %s' % (zone, newzone))
        zone = newzone
    tz = pytz.timezone(zone)


Contribute
==========

- Issue Tracker: https://github.com/uwosh/uwosh.oie.studyabroadstudent/issues
- Source Code: https://github.com/uwosh/uwosh.oie.studyabroadstudent
- Documentation: https://github.com/uwosh/uwosh.oie.studyabroadstudent/docs


Credits
=======

The project was paid for by the Office of International Education at the University of Wisconsin Oshkosh (http://uwosh.edu/oie).

Implementation by T. Kim Nguyen at Wildcard Corp. (https://wildcardcorp.com).


Support
=======

If you are having issues, please let us know via the GitHub issue tracker.



License
-------

The project is licensed under the GPLv2.


CHANGES.rst
Changelog
=========


2.0.4 (unreleased)
------------------

- Nothing changed yet.


2.0.3 (2021-12-03)
------------------

- omit correct fields for participant view and edit forms
- fix needless abbreviation, fix email template
- add new forms / organize control panel
- control panel schema refresh upgrade step
- remove old uneeded json files


2.0.2 (2021-11-10)
------------------

- Added IntNoFormatFieldWidget to not format years (ex: 2,021 should be 2021)
- changed participant graduationYear to use this widget
- constrained graduationYear to CURRENT_YEAR+10


2.0.1 (2021-11-08)
------------------

- rename transition 'omit-special-student-status-application' to 'do-not-require-a-special-student-status-application'
- add the exit-transitions 'pending-special-student-status-graduate' and 'do-not-require-a-special-student-status-application' to the state 'checking-for-special-student-status'
- remove the exit transition 'apply-for-special-student-status-graduate' from the state 'pending-special-student-status-undergraduate'
- remove the exit transition 'omit-special-student-status-application' from the state 'pending-special-student-status-graduate'


2.0.0 (2021-10-21)
------------------

- tweak state transition error message
- alterations and additions to participant workflow
- upgrade stuff for 1008


1.0.18 (2021-10-14)
-------------------

- Fixed airport_transfer typo


1.0.17 (2021-09-23)
-------------------

- change submit message


1.0.16 (2021-09-22)
-------------------

- bump metadata.xml


1.0.15 (2021-09-22)
-------------------

- adjust participant view action
- adjust show toolbar permission


1.0.14 (2021-09-22)
-------------------

- Changes to download form link rendering
- minor role permission change


1.0.13 (2021-09-17)
-------------------

- serve minimized react


1.0.12 (2021-09-17)
-------------------

- Fix Participant permissions
- Get functioning ux for applying with login and register
- add my applications portal tab functionality
- bring portal tabs into code


1.0.11 (2021-09-13)
-------------------

- Discover programs view changes to simplify semantics and use css grid


1.0.10 (2021-09-13)
-------------------

- changes to allow unauthenticated users to view necessary info associated with,
  but not directly contained within discoverable programs (via a uid linking to another object)
- start using get_object_from_uid
- refactor discover programs form to fetch data from endpoint
- discover programs paging tweak
- various and multiple formatting changes
- upgrade step to reindex programs
- reset currentPage of discover programs form if out of bounds after updating filter


1.0.9 (2021-09-10)
------------------

- additional manager-role permission for associated program view content items
- with_manager_permissions util decorator


1.0.8 (2021-09-09)
------------------

- fix permissions for loading programsearch details while anonymous
- fix programview.pt to handle missing attributes


1.0.7 (2021-09-09)
------------------

- fix programsearch typo


1.0.6 (2021-09-09)
------------------

- change program_leader on discoverprograms to be a last name instead of UID
- remove pt reference to folder_listing for cooperatingpartnerview and reformat
- change default site title


1.0.5 (2021-09-08)
------------------

- Made permission changes for programs in discoverable state to be able to be viewed anonymously
- Upgrade step and create anonymous user for impersonator


1.0.4 (2021-08-27)
------------------

- Second deployment update
- added OIEHomePage to show /discover on root
- made changes to discover programs filtering and rewrote programsearch.js
- changes in some required fields and add STATES_FOR_DISPLAYING_PROGRAMS


1.0.0 (2021-08-06)
------------------

- Initial release.
  [tkimnguyen]


src/uwosh/oie/studyabroadstudent/constants.py
from datetime import date

CURRENT_YEAR = date.today().year
NEXT_YEAR = CURRENT_YEAR + 1

DURATIONS = [
    7,
    14,
    21,
    35,
    65,
    95,
    125,
    155,
    185,
    215,
    245,
    275,
    305,
    335,
    365,
]
PASSWORD_LENGTH = 50
SPECIAL_DAYS = [
    'first_day_of_spring_semester_classes',
    'last_day_of_spring_semester_classes',
    'first_day_of_spring_interim_classes',
    'last_day_of_spring_interim_classes',
    'official_spring_graduation_date',
    'first_day_of_summer_i_classes', 'last_day_of_summer_i_classes',
    'first_day_of_summer_ii_classes',
    'last_day_of_summer_ii_classes',
    'official_summer_graduation_date',
    'first_day_of_fall_semester_classes',
    'last_day_of_fall_semester_classes',
    'first_day_of_winter_interim_classes',
    'last_day_of_winter_interim_classes',
    'official_fall_graduation_date',
    'spring_interim_summer_fall_semester_participant_orientation_deadline',
    'spring_interim_summer_fall_semester_in_person_orientation',
    'winter_interim_spring_semester_participant_orientation_deadline',
    'winter_interim_spring_semester_in_person_orientation',
    'spring_interim_summer_fall_semester_payment_deadline_1',
    'spring_interim_payment_deadline_2',
    'summer_payment_deadline_2', 'fall_semester_payment_deadline_2',
    'winter_interim_spring_payment_deadline_1',
    'winter_interim_spring_payment_deadline_2',
    'request_for_proposals_due_date',
]
EMERGENCY_EMAIL_FIELD_DESCRIPTION = (
    'Strongly recommended.  By typing in an email address here, you permit the UWO OIE to send '
    'non-emergency messages intended to update contacts about significant unanticipated events that '
    'have occurred or may occur and which have involved or may involve an increase in program risk.'
)
EMERGENCY_PHONE_SECONDARY_DESCRIPTION = (
    'Include area code (and country code if the phone does not have a U.S. phone number).'
)
EMERGENCY_PHONE_PRIMARY_DESCRIPTION = f'Strongly recommended. {EMERGENCY_PHONE_SECONDARY_DESCRIPTION}'

DEFAULT_ELIGIBILITY_REQUIREMENT_1 = (
    'Undergraduate students must have a minimum G.P.A. of 2.0, be in good standing, and meet course '
    'prerequisites. As part of the application process, your disciplinary file will be reviewed. '
    'If you have a campus disciplinary file, this will be taken into consideration when determining '
    'your eligibility. Undergraduates who do not meet minimum GPA and course requirements should contact '
    'the program leader prior to applying to determine course eligibility.'
)
DEFAULT_ELIGIBILITY_REQUIREMENT_2 = (
    'Undergraduate students must have a minimum G.P.A. of 2.0, be in good standing, and meet course '
    'prerequisites. Undergraduate students who are required to complete a Quest III course to fulfill '
    'graduation requirements may apply. Students must have a minimum G.P.A. of 2.0, be in good standing, '
    'meet course prerequisites, and meet Quest III prerequisites. As part of the application process, '
    'your disciplinary file will be reviewed. If you have a campus disciplinary file, this will be taken '
    'into consideration when determining your eligibility. Undergraduates who are not required to complete '
    'a Quest III course to graduate are not eligible; however, they may apply for consideration on a '
    'space-available basis along with USP & instructor consent. Seats will be assigned to these applicants '
    'only after all applicants who are required to complete Quest III to graduate and who have met the '
    'STEP III deadline have been awarded a seat.'
)
DEFAULT_ELIGIBILITY_REQUIREMENT_3 = (
    'Students must have a minimum G.P.A. of 3.25, be accepted into the Professional Counseling Graduate '
    'program, be in good standing, and meet course prerequisites. As part of the application process, '
    'your disciplinary file will be reviewed. If you have a campus disciplinary file, this will be taken '
    'into consideration when determining your eligibility. Undergraduates who do not meet minimum GPA '
    'and course requirements should contact the program leader prior to applying to determine course '
    'eligibility.'
)
DEFAULT_ELIGIBILITY_REQUIREMENTS = (
    DEFAULT_ELIGIBILITY_REQUIREMENT_1,
    DEFAULT_ELIGIBILITY_REQUIREMENT_2,
    DEFAULT_ELIGIBILITY_REQUIREMENT_3,
)
DEADLINES = {
    'steps_1_and_2': {
        'spring interim': 'the 1st Friday of Spring Semester',
        'summer semester': 'the last Friday in February',
        'fall semester': 'the last Friday in February',
        'fall interim': 'the 2nd Friday in September',
        'spring semester': 'the last Friday in September',
        'base_description': (
            'The STEPs I & II application deadline must be the OIE default student application date, '
            'or a date that is two weeks prior to the contracted date to release airline tickets, '
            'whichever is earlier. Alternatively, the Program Liaison may identify an even earlier '
            'deadline, provided OIE can provide sufficient staffing in the week leading up to the '
            'proposed deadline.'
        ),
    },
    'step_3': {
        'spring interim': 'the 2nd Friday of Spring Semester',
        'summer semester': 'the 1st Friday in March',
        'fall semester': 'the 1st Friday in March',
        'fall interim': 'the 3rd Friday in September',
        'spring semester': 'the 1st Friday in October',
        'base_description': (
            'The STEP III application deadline must be the OIE default student application date, '
            'or a date that is one week prior to the contracted date to release airline tickets, '
            'whichever is earlier. Alternatively, the Program Liaison may identify an even earlier '
            'deadline, provided OIE can provide sufficient staffing in the week leading up to the '
            'proposed deadline.'
        ),
    },
    'step_4': {
        'base_description': (
            'The STEP IV application deadline must take into consideration external deadlines and '
            'processing time in the OIE from the point of receiving completed application documents, '
            'and the anticipated dates on which documents can be sent to external partners and '
            'received by them.'
        ),
    },
}
DEFAULT_TUITION_AND_FEES = (
    '$0 Tuition; $0 Seg Fees',
    '$0 Tuition; $0 Seg Fees; relevant tuition equivalent to be charged',
    'No Change to Course Tuition & Fees',
    'Tuition Charged for 01 cr + related fees; charge outside of the tuition plateau',
    'Tuition Charged for 01 cr; Seg Fees Also Charged',
    'Tuition Charged for 02 cr; Seg Fees Also Charged',
    'Tuition Charged for 03 cr; $0 Seg Fees',
    'Tuition Charged for 03 cr + related fees; charge outside of the tuition plateau',
    'Tuition Charged for 03 cr; Seg Fees Also Charged',
    'Tuition Charged for 04 cr; Seg Fees Also Charged',
    'Tuition Charged for 05 cr; Seg Fees Also Charged',
    'Tuition Charged for 06 cr outside of the tuition plateau; $0 Seg Fees',
    'Tuition Charged for 06 cr; $0 Seg Fees',
    'Tuition Charged for 06 cr; Seg Fees Also Charged',
    'Tuition Charged for 07 cr; Seg Fees Also Charged',
    'Tuition Charged for 08 cr; Seg Fees Also Charged',
    'Tuition Charged for 09 cr; Seg Fees Also Charged',
    'Tuition Charged for 1.5 cr; Seg Fees Also Charged',
    'Tuition Charged for 10 cr; Seg Fees Also Charged',
    'Tuition Charged for 11 cr; Seg Fees Also Charged',
    'Tuition Charged for 12 cr; $0 Seg Fees',
    'Tuition Charged for 12 cr; Seg Fees Also Charged',
)


CALENDAR_YEAR_DESCRIPTION = (
    'Select the calendar year during which the program will run. '
    'This is not the year associated with the term of study. '
    f'For example, a January interim program running from Jan 2-28, {NEXT_YEAR} '
    f'will be associated with "{NEXT_YEAR}". A program running Dec 28, {CURRENT_YEAR}-Jan 28, {NEXT_YEAR} '
    f'will also be associated with "{NEXT_YEAR}".'
)

STATES_FOR_DISPLAYING_PROGRAMS = (
    'pending-discussions-with-program-manager',
    'request-for-proposals-under-development',
    'request-for-proposals-under-liaison-review',
    'pending-provider-responses',
    'provider-proposals-under-oie-review',
    'provider-proposals-under-liaison-review',
    'pending-program-fee-determination-by-oie',
    'program-fee-under-liaison-review',
    'program-fee-pending-publication',
    'applicants-considering-change',
    'application-intake-in-progress',
)
